CASSANDRA-19340 - Support writing UDTs

JeetKunDoug opened the pull request:
This commit brings support for UDTs to the Bulk Writer. It converts Spark nested StructType(s) into UDT values.

JeetKunDoug committed:
Make sure bridge exists

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkWriterContext.java:
```diff
@@ -64,7 +65,6 @@ private CassandraBulkWriterContext(@NotNull BulkSparkConf conf,
         this.conf = conf;
         this.clusterInfo = clusterInfo;
         String lowestCassandraVersion = clusterInfo.getLowestCassandraVersion();
-        CassandraBridge bridge = CassandraBridgeFactory.get(lowestCassandraVersion);
```
Nothing wrong still initializing the bridge here. Should we preserve it?

JeetKunDoug replied:
Mostly a matter of getting the lowest version as late as possible - imagine a mid-upgrade scenario where a node comes back online with a newer version (and it was the last one to be upgraded)... In the grand scheme of things, this is pretty unlikely to happen, and the previous version's libraries will still work, but it seemed worthwhile to allow it to be lazy.

yifan-c replied:
So, as you said, the effect is the same for both eager or lazy instantiation. But you can avoid synchronization if instantiate eagerly.

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkWriterContext.java:
```diff
@@ -87,11 +87,21 @@ private CassandraBulkWriterContext(@NotNull BulkSparkConf conf,
+    {
+        if (this.bridge == null)
+        {
+            this.bridge = CassandraBridgeFactory.get(this.clusterInfo.getLowestCassandraVersion());
```
it feels like the driver should perform this operation once (it's a potentially expensive operation) and we should serialize the resulting version value to the executors. Executors should just use the version value that the driver computed.

JeetKunDoug replied:
Same comment as above - it's possible the "lowest version" has changed from the beginning of the job. Again, maybe not worth the extra load but I don't think it's particularly heavy-weight.

frankgh replied:
On a smaller cluster maybe not, but I think it can add up on a big Cassandra cluster when running a large spark job. I think there's a tradeoff here. In the case where that one instance coming back with a lower version, the instance would probably miss the write (as it would be unable to import the SStables) and the data would need to be repaired. So the question is whether we want so save some REST endpoint requests on all the job runs vs save on repair cost on a very unlikely scenario that would very rarely happen.

yifan-c replied:
It is safe (or safer) to use the low version among the entire job runtime, which is essential the low version when the job starts. 
Cassandra can take the sstable written with low version. 
There could be a risk that a node has to downgrade and the written sstables are in the new version. The scenario is rare, but you get my point.

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CqlTableInfoProvider.java:
```diff
@@ -98,9 +98,20 @@ public CqlField.CqlType getColumnType(String columnName)
-                       .collect(Collectors.toList());
+        List<ColumnType<?>> types = cqlTable.partitionKeys().stream()
+                                              .map(cqlField -> {
+                                                  String typeName = cqlField.type().cqlName().toLowerCase();
```
why do we lower case the cqlName? Can we add tests for quoted UDTs as well?

JeetKunDoug replied:
Note that this map (`DATA_TYPES`) currently doesn't support UDTs at all - it's something we need to implement separately to support UDTs for partition/clustering key columns (as `Tokenizer` doesn't support them at this time). So, for now, we throw if we can't find the type in that map, but we'll need to update the logic once the Tokenizer supports them.

frankgh reviewed (commented):
Not a full review yet, but I will push my review so far so it doesn't get lost

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraSchemaInfo.java:
```diff
@@ -19,19 +19,33 @@
     private static final long serialVersionUID = -2327383232935001862L;
     private final TableSchema tableSchema;
+    private final Set<String> userDefinedTypeStatements;
+    private final CqlTable cqlTable;
```
variable not exposed and not used anywhere

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -95,6 +105,7 @@ public RecordWriter(BulkWriterContext writerContext, String[] columnNames)
         this.writeValidator = new BulkWriteValidator(writerContext, failureHandler);
         this.digestAlgorithm = this.writerContext.job().digestAlgorithmSupplier().get();
 
+
```
nit: remove the new empty line

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:
```diff
@@ -108,6 +119,26 @@ private String getStreamId(TaskContext taskContext)
         return String.format("%d-%s", taskContext.partitionId(), UUID.randomUUID());
     }
 
+    private CqlTable cqlTable()
```
The only callsite of `cqlTable()` is already `synchronized` on record writer. I think you can remove the double-checked locking entirely, if keeping the synchronization at `getUdt()`.

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SqlToCqlTypeConverter.java:
```diff
@@ -763,6 +766,12 @@ else if (object instanceof Map)
+        @Override
+        public String toString()
+        {
+            return "Map";
```
How about the richer string? 

```suggestion
            return "Map<" + keyConverter.toString() + ", " + valConverter.toString() + '>';
```

JeetKunDoug replied:
Ha - sure... although technically this PR didn't actually change that, it's just a formatting thing

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SqlToCqlTypeConverter.java:
```diff
@@ -788,11 +797,54 @@ else if (object instanceof scala.Tuple2)
+    public static class UdtConverter extends NullableConverter<Object>
+    {
+        private final String name;
+        private final HashMap<Object, Converter<?>> converters;
```
Key is `String`

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SqlToCqlTypeConverter.java:
```diff
@@ -788,11 +797,54 @@ else if (object instanceof scala.Tuple2)
+    {
+        private final String name;
+        private final HashMap<Object, Converter<?>> converters;
+        private CqlField.CqlUdt udt;
```
Unused.

yifan-c commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/SqlToCqlTypeConverter.java:
```diff
@@ -788,11 +797,54 @@ else if (object instanceof scala.Tuple2)
         }
+    }
+
+    public static class UdtConverter extends NullableConverter<Object>
```
Why it converts to `Object` instead of `BridgeUdtValue`?

JeetKunDoug replied:
Originally, I was trying to convert to the final values, but it ended up it wasn't possible due to some serialization issues, so it returns the `BridgeUdtValue` - I'll update it to be more specific.

yifan-c commented on cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/BulkWriteUdtTest.java:
```diff
@@ -0,0 +1,208 @@
+                                 .containsExactlyInAnyOrderElementsOf(sourceEntries);
+    }
+
+    private static String getFormatedSourceEntry(Row row)
```
typo: `getFormattedSourceEntry` (double t)

yifan-c commented on cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/TableSchemaNormalizeTest.java:
```diff
@@ -296,6 +303,24 @@ public void testNestedNormalization()
+                                .add(new StructField("f2", DataTypes.StringType, false, Metadata.empty()));
+
+        GenericRowWithSchema source = new GenericRowWithSchema(new Object[]{1, "course"}, structType);
+        // NOTE: UDT Types cary their type name around, so the use of `udt_field` consistently here is a bit
```
`carry`?

yifan-c commented on cassandra-four-zero/src/main/java/org/apache/cassandra/bridge/SSTableWriterImplementation.java:
```diff
@@ -81,17 +85,23 @@ static CQLSSTableWriter.Builder configureBuilder(String inDirectory,
-               // see org.apache.cassandra.spark.bulkwriter.CassandraBulkSourceRelation.insert
-               .sorted()
-               .withMaxSSTableSizeInMiB(bufferSizeMB);
+        CQLSSTableWriter.Builder builder = CQLSSTableWriter
+                                           .builder()
+                                           .inDirectory(inDirectory)
+                                           .forTable(createStatement)
+                                           .withPartitioner(cassPartitioner)
+                                           .using(insertStatement)
+                                           // The data frame to write is always sorted,
+                                           // see org.apache.cassandra.spark.bulkwriter.CassandraBulkSourceRelation.insert
+                                           .sorted()
+                                           .withMaxSSTableSizeInMiB(bufferSizeMB);
+        for (String udt : udts)
+        {
+            builder.withType(udt);
+        }
+        return builder;
```
nit
```suggestion
        CQLSSTableWriter.Builder builder = CQLSSTableWriter.builder();
        for (String udt : udts)
        {
            builder.withType(udt);
        }
        
        return builder
               .inDirectory(inDirectory)
               .forTable(createStatement)
               .withPartitioner(cassPartitioner)
               .using(insertStatement)
               // The data frame to write is always sorted,
               // see org.apache.cassandra.spark.bulkwriter.CassandraBulkSourceRelation.insert
               .sorted()
               .withMaxSSTableSizeInMiB(bufferSizeMB);
```

JeetKunDoug replied:
I'm not sure why this is any different/better than the existing code - just less indented?

yifan-c reviewed (commented):
mostly nits for cosmetic changes

frankgh commented on cassandra-analytics-core/src/test/java/org/apache/cassandra/spark/bulkwriter/MockBulkWriterContext.java:
```diff
@@ -77,6 +79,7 @@ public class MockBulkWriterContext implements BulkWriterContext, ClusterInfo, Jo
     private ConsistencyLevel.CL consistencyLevel;
     private int sstableDataSizeInMB = 128;
     private int sstableWriteBatchSize = 2;
+    private CassandraBridge bridge = CassandraBridgeFactory.get(CassandraVersion.FOURZERO);
```
should this be passed to the mock BW context? I wonder what tests for 50, 51 will look like if we use 40 here always?

frankgh commented on cassandra-bridge/src/main/java/org/apache/cassandra/spark/data/BridgeUdtValue.java:
```diff
@@ -0,0 +1,69 @@
+        return Objects.hash(name, udtMap);
+    }
+
+    public String toString()
```
NIT, override annotation

frankgh commented on cassandra-bridge/src/main/java/org/apache/cassandra/spark/data/BridgeUdtValue.java:
```diff
@@ -0,0 +1,69 @@
+        return Objects.equals(name, udtValue.name) && Objects.equals(udtMap, udtValue.udtMap);
+    }
+
+    public int hashCode()
```
NIT, override annotation

frankgh commented on cassandra-bridge/src/main/java/org/apache/cassandra/spark/data/BridgeUdtValue.java:
```diff
@@ -0,0 +1,69 @@
+        this.udtMap = valueMap;
+    }
+
+    public boolean equals(Object o)
```
NIT, override annotation

JeetKunDoug commented:
Rebased, addressed comments, green CircleCI: https://app.circleci.com/pipelines/github/JeetKunDoug/cassandra-analytics/141/workflows/9d8c886c-81c9-4d43-9e78-43a52fab14f9

JeetKunDoug commented:
Rebased/squashed/fixed CHANGES.txt conflict (no other changes/conflicts)

frankgh commented on cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/CassandraBulkWriterContext.java:
```diff
@@ -92,11 +92,17 @@ protected CassandraBulkWriterContext(@NotNull BulkSparkConf conf,
+    @Override
+    public CassandraBridge bridge()
+    {
+        return this.bridge;
```
if this is serialized to executors, and it's declared as transient, we'll need to initialize it for the executors at some point. I suggest we add a field called `private final String lowestCassandraVersion;` and we initialize it in the constructor (line 72). I think at the executor level we'll see NPEs for the bridge

frankgh reviewed (commented):
I've added one additional comment that I think needs to be addressed

frankgh reviewed (approved):
+1 Looks good. Thanks
